pkg/virt-controller/services/template.go

```diff
@@ -600,10 +610,29 @@ func (t *templateService) RenderLaunchManifest(vmi *v1.VirtualMachineInstance) (
 			},
 		})
 
+		reqMemDiff := resource.NewScaledQuantity(0, resource.Kilo)
+		limMemDiff := resource.NewScaledQuantity(0, resource.Kilo)
+		// In case the guest memory and the requested memeory are diffrent, add the difference
+		// to the to the overhead
+		if vmi.Spec.Domain.Memory != nil && vmi.Spec.Domain.Memory.Guest != nil {
+			requests := vmi.Spec.Domain.Resources.Requests.Memory().Value()
+			limits := vmi.Spec.Domain.Resources.Limits.Memory().Value()
+			guest := vmi.Spec.Domain.Memory.Guest.Value()
+			if requests > guest {
+				reqMemDiff.Add(*vmi.Spec.Domain.Resources.Requests.Memory())
+				reqMemDiff.Sub(*vmi.Spec.Domain.Memory.Guest)
+			}
+			if limits > guest {
+				limMemDiff.Add(*vmi.Spec.Domain.Resources.Limits.Memory())
+				limMemDiff.Sub(*vmi.Spec.Domain.Memory.Guest)
+			}
+		}
 		// Set requested memory equals to overhead memory
-		resources.Requests[k8sv1.ResourceMemory] = *memoryOverhead
+		reqMemDiff.Add(*memoryOverhead)
```

Comment:
Then it is not a diff anymore. Could you use a better name for the variable to make it clear that it is total memory consisting of the request of VM+overhead?

Reply:
@phoracek - That's not a guest memory + overhead.

In general, when we start a VMI we need to request more memory for the compute container than the guest needs.
We refer to the difference between the container memory and the guest memory as a memory overhead[1].
The overhead calculation includes the memory needed for qemu threads, graphic card memory, os/code running in the compute container and more... see [1]

When it comes to huge pages, we need to request k8s to allocate some of the memory with specific page size (`resources.Requests[hugepageType]`) and the rest of the memory (overhead + difference between guest memory and the memory requested for the container) with a default page size (`resources.Requests[k8sv1.ResourceMemory]`)
The memory allocated for huge pages will be provided to the guest for it to use as it's memory.

In this PR, we allow the guest memory `vmi.Spec.Domain.Memory.Guest` to be different from `vmi.Spec.Domain.Resources.Requests` therefore, we need to allocate huge pages according to the guest's memory (`vmi.Spec.Domain.Memory.Guest`) and the difference (requested mem - guest memory + overhead) request to allocate as a memory with a default page size (`resources.Requests[k8sv1.ResourceMemory]`).  

[1] https://github.com/kubevirt/kubevirt/blob/master/pkg/virt-controller/services/template.go#L1069

Reply:
Thanks.

Reply:
I get it now.

Maybe it is just my ignorance and lack of knowledge about kubevirt, but the fact that there is no design document for this (is there) anyone new to kubevirt can read on is frightening. It makes is really hard for somebody new to start contributing to kubevirt even with reviews.